File: src/core/effects/temporalGhosts.ts
/**
 * Temporal ghosts — the shared shape behind every effect that draws the layer
 * at OTHER points in time.
 *
 * Echo and Wide Time differ only in which times they pick and how bright each
 * copy is. The emission itself — resample the transform at t+dt, rebuild the
 * 3D matrix, strip the matte/motion-blur/adjustment flags, composite — is
 * identical, and it is forty lines of subtle code inside `buildSnapshot`. This
 * module is the part that differs, so the part that does not can stay written
 * once. `CompositionPass` already paid for the alternative: two copies of the
 * layer-effects chain, one of which silently erased any effect type it did not
 * know (see gotcha-motion-two-effect-chains).
 *
 * Everything here is a pure function of the animation and the clock — no frame
 * history — which is what keeps these effects scrub-stable and cacheable. That
 * is also the ceiling: AE's Time Displacement, Time Difference and Pixel Motion
 * Blur read previously RENDERED frames and cannot be expressed this way.
 */

import type { Effect } from './effects';
import { effectNumber } from './effects';
import { readEchoConfig, type EchoOperator } from './echo';

/** One ghost copy: how far from now, and how strongly it draws. */
export interface GhostStep {
  /** Seconds relative to the current frame. Negative = into the past. */
  dt: number;
  /** Opacity multiplier applied on top of the layer's own, 0..1. */
  opacity: number;
}

export interface GhostSpec {
  steps: GhostStep[];
  /** Blend mode the copies composite with. */
  blend: EchoOperator;
  /** True when the copies draw OVER the current frame. */
  inFront: boolean;
}

/** Ghosts too faint to see cost a full layer draw each. */
const MIN_VISIBLE = 0.002;

/**
 * Wide Time — AE's CC Wide Time: copies from BOTH directions, evenly weighted,
 * which reads as a temporal smear rather than as a trailing wake.
 *
 * Steps are whole frames because that is the unit AE's controls are in, so this
 * needs the comp's frame rate. Weighting is `1/(total + 1)` — the current frame
 * counts as one of the copies, so a 2-forward/2-back setting averages five
 * moments rather than blowing out to 5× brightness.
 */
function readWideTime(e: Effect, fps: number): GhostSpec | null {
  const fwd = Math.max(0, Math.min(64, Math.round(effectNumber(e, 'forwardSteps'))));
  const back = Math.max(0, Math.min(64, Math.round(effectNumber(e, 'backwardSteps'))));
  if (fwd + back === 0 || fps <= 0) return null;
  const weight = 1 / (fwd + back + 1);
  const steps: GhostStep[] = [];
  for (let k = 1; k <= back; k++) steps.push({ dt: -k / fps, opacity: weight });
  for (let k = 1; k <= fwd; k++) steps.push({ dt: k / fps, opacity: weight });
  // `normal` and behind: this is an average of moments, and adding them would
  // make a static layer brighter than it is.
  return { steps, blend: 'normal', inFront: false };
}

/** Echo's steps: a decaying run in ONE direction, the sign of `time`. */
function echoSteps(cfg: NonNullable<ReturnType<typeof readEchoConfig>>): GhostStep[] {
  const steps: GhostStep[] = [];
  for (let k = 1; k <= cfg.count; k++) {
    steps.push({ dt: k * cfg.time, opacity: cfg.startIntensity * Math.pow(cfg.decay, k - 1) });
  }
  return steps;
}

/**
 * The ghost copies this layer's effect stack asks for, or null for none.
 *
 * Echo wins when a layer carries both. They are the same mechanism pointed at
 * different times, and compounding them would emit `echo.count × wide.steps`
 * layers — a combinatorial cost from a stack the user would read as "two
 * effects", plus a picture neither effect describes.
 *
 * Ordered farthest-first, so nearer copies paint over more distant ones and the
 * current frame lands on top of all of them.
 */
export function readGhostSpec(effects: ReadonlyArray<Effect>, fps: number): GhostSpec | null {
  const echo = readEchoConfig(effects);
  if (echo && echo.count > 0 && echo.time !== 0) {
    return { steps: order(echoSteps(echo)), blend: echo.operator, inFront: echo.echoesInFront };
  }
  const wide = effects.find((x) => x.type === 'wide-time' && x.enabled !== false);
  if (wide) {
    const spec = readWideTime(wide, fps);
    if (spec) return { ...spec, steps: order(spec.steps) };
  }
  return null;
}

function order(steps: GhostStep[]): GhostStep[] {
  return steps
    .filter((s) => s.opacity > MIN_VISIBLE)
    .sort((a, b) => Math.abs(b.dt) - Math.abs(a.dt));
}

File: native/engine/src/scene/temporal_ghosts.cpp
#include "temporal_ghosts.hpp"

#include <algorithm>
#include <array>
#include <cmath>

#include "effects_port.hpp"
#include "jsmath.hpp"

namespace premation::scene {
namespace {

/// Ghosts too faint to see cost a full layer draw each (MIN_VISIBLE).
constexpr double kMinVisible = 0.002;

struct EchoOp {
  std::string_view blend;
  bool inFront;
};
/// ECHO_OPERATORS, menu order.
constexpr std::array<EchoOp, 6> kEchoOps = {{
    {"add", false}, {"lighten", false}, {"darken", false}, {"screen", false}, {"normal", false}, {"normal", true},
}};

double clamp01(double v) { return std::max(0.0, std::min(1.0, v)); }

const Json* find_enabled(const std::vector<Json>& effects, std::string_view type) {
  for (const Json& e : effects) {
    if (e.at("type").is_string() && e.at("type").str() == type && !(e.at("enabled").is_bool() && !e.at("enabled").b())) {
      return &e;
    }
  }
  return nullptr;
}

/// order(): drop the invisible, farthest first (Array.prototype.sort is stable).
std::vector<GhostStep> order(std::vector<GhostStep> steps) {
  std::erase_if(steps, [](const GhostStep& s) { return !(s.opacity > kMinVisible); });
  std::ranges::stable_sort(steps, [](const GhostStep& a, const GhostStep& b) { return std::abs(b.dt) < std::abs(a.dt); });
  return steps;
}

}  // namespace

std::optional<GhostSpec> read_ghost_spec(const std::vector<Json>& effects, double fps) {
  if (const Json* e = find_enabled(effects, "echo")) {
    // readEchoConfig: an operator index naming no operator falls back to Add.
    const double opIdx = motion::js::round(effect_number(*e, "echoOperator"));
    const EchoOp op = opIdx >= 0 && opIdx < static_cast<double>(kEchoOps.size()) ? kEchoOps[static_cast<std::size_t>(opIdx)]
                                                                                  : kEchoOps[0];
    const double time = effect_number(*e, "echoTime");
    const double count = std::max(0.0, std::min(64.0, motion::js::round(effect_number(*e, "numEchoes"))));
    const double start = clamp01(effect_number(*e, "startIntensity") / 100);
    const double decay = clamp01(effect_number(*e, "decay") / 100);
    if (count > 0 && time != 0) {
      GhostSpec g;
      std::vector<GhostStep> steps;
      for (int k = 1; k <= static_cast<int>(count); ++k) steps.push_back({k * time, start * motion::js::pow(decay, k - 1)});
      g.steps = order(std::move(steps));
      g.blend = std::string(op.blend);
      g.inFront = op.inFront;
      return g;
    }
  }
  if (const Json* e = find_enabled(effects, "wide-time")) {
    const double fwd = std::max(0.0, std::min(64.0, motion::js::round(effect_number(*e, "forwardSteps"))));
    const double back = std::max(0.0, std::min(64.0, motion::js::round(effect_number(*e, "backwardSteps"))));
    if (fwd + back == 0 || fps <= 0) return std::nullopt;
    const double weight = 1 / (fwd + back + 1);
    std::vector<GhostStep> steps;
    for (int k = 1; k <= static_cast<int>(back); ++k) steps.push_back({-k / fps, weight});
    for (int k = 1; k <= static_cast<int>(fwd); ++k) steps.push_back({k / fps, weight});
    GhostSpec g;
    g.steps = order(std::move(steps));
    return g;
  }
  return std::nullopt;
}

std::vector<RLayer> ghost_layers(const RLayer& layer, const GhostSpec& spec, double t, const GhostSampler& sample,
                                 double localX, double localY, double localRot, double px, double py, double rot,
                                 const GhostPlace3D& place3d) {
  std::vector<RLayer> out;
  for (std::size_t k = 0; k < spec.steps.size(); ++k) {
    const GhostStep& step = spec.steps[k];
    const double ti = t + step.dt;
    if (ti < 0) continue;
    const double op = layer.opacity * step.opacity;
    if (op <= 0.002) continue;
    RLayer g = layer;
    g.id = layer.id + "__echo" + std::to_string(k);
    g.opacity = op;
    g.blend = spec.blend;
    g.matte = std::nullopt;
    g.matteSourceId = std::nullopt;
    g.isMatteSource = false;
    g.isAdjustment = false;
    g.motionSamples.clear();
    const double dx = sample("x", ti).value_or(localX) - localX;
    const double dy = sample("y", ti).value_or(localY) - localY;
    const double drot = sample("rotation", ti).value_or(localRot) - localRot;
    if (place3d) {
      place3d(g, ti, dx, dy, drot);
    } else {
      g.x = px + dx;
      g.y = py + dy;
      g.rotation = rot + drot;
    }
    out.push_back(std::move(g));
  }
  return out;
}

}  // namespace premation::scene
